### ISSUE TITLE
Typescript should not allow to add

### ISSUE THREAD

**Issue by LogvinovLeon** [Microsoft/TypeScript#21230]
**TypeScript Version:**  2.6

**Code**

```ts
let str = "abacaba";
str = str + 1;
```
https://www.typescriptlang.org/play/#src=let%20str%20%3D%20%22abacaba%22%3B%0D%0Astr%20%3D%20str%20%2B%201%3B%0D%0A%2F%2F%20keck

**Expected behavior:**
Should throw an error because type `number` should not be addable to a type `string`.

**Actual behavior:**
Compiles without any issues and causes an implicit type conversion

Maybe that's an intended design decision. In that case I'd suggest it to be disabled in s strict mode. This is not the type of behaviour one would expect from a strictly typed language.

**Comment by NaridaL** [comment]
You can use https://palantir.github.io/tslint/rules/restrict-plus-operands/

**Comment by LogvinovLeon** [comment]
Thanks. Will do!
